apis/placement/v1beta1/overriddenresourcesnapshot_types.go

```diff
@@ -0,0 +1,90 @@
+// target cluster.
+// Once the snapshot has been overridden by the system, the manifest won't be updated because of the cluster label change (e.g
+// the override rule are not apply to the target cluster anymore). To make the changes effective, you need to update the
+// resource or its related override.
+//
+// Its spec is immutable.
+// We may need to produce more than one overriddenClusterResourceSnapshot for all the resources a placement selected to
+// get around the 1MB size limit of k8s objects.
+// The naming convention of a overriddenClusterResourceSnapshot is {clusterResourceSnapshotName}-{targetCluster}-{random-suffix}.
+// Each snapshot MUST have the following labels:
+//   - `CRPTrackingLabel` which points to its owner CRP.
+//   - `MasterSnapshotNameLabel` which is the name of the master overriddenClusterResourceSnapshot in a group.
+//   - `ClusterResourceSnapshotNameLabel` which is the name of the clusterResourceSnapshot.
+//   - `TargetClusterLabel` which is the overridden result of the clusterResourceSnapshot on the target cluster.
+//   - `IsLatestSnapshotLabel` which indicates whether the snapshot is the latest one.
+//
+// All the snapshots within the same index group must have the same ResourceIndexLabel.
+type OverriddenClusterResourceSnapshot struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of OverriddenResourceSnapshot.
+	// +required
+	Spec OverriddenResourceSnapshotSpec `json:"spec"`
+
+	// The observed status of ResourceSnapshot.
+	// +optional
+	Status OverriddenResourceSnapshotStatus `json:"status,omitempty"`
+}
+
+type OverriddenResourceSnapshotSpec struct {
+	TargetCluster string `json:"targetCluster"`
+
+	ClusterResourceSnapshotName string `json:"clusterResourceSnapshotName"`
+
+	ResourceOverrides []OverrideSnapshotIdentifier `json:"resourceOverrides"`
+
+	// ResourceOverridesHash is the sha-256 hash value of the ResourceOverrides field.
+	// +required
+	ResourceOverridesHash []byte `json:"resourceOverridesHash"`
+}
+
+type OverriddenResourceSnapshotStatus struct {
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+
+	// Conditions is an array of current observed conditions for OverriddenResourceSnapshot.
+	// The type is Overridden to indicate whether the overridden of the resources on the target cluster is valid or not.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// NumberOfSnapshots indicates how many snapshots are created within the same group and generated from the same clusterResourceSnapshot
+	// and its overrides.
+	// +optional
+	// +kubebuilder:default=1
+	NumberOfSnapshots int `json:"numberOfSnapshots,omitempty"`
+
+	// Workload represents the manifest workload to be deployed on spoke cluster.
+	Workload WorkloadTemplate `json:"workload,omitempty"`
```

Comment:
AFAIK, most of the fields in the status are actually known when the CR object is created instead of having another controller reconcile based on the spec. Hence, I wonder why those are in the status? I can see condition be part of status just to keep the convention.

Reply:
Or should we actually use the work generator to reconcile this object which makes the flow a lot more natural? I know the problem is that we won't be able to "keep the last known good state" but I am not sure if that's what a state driven system normally guarantees.